New test failures du to numeircal precision

Some tests added in #1770 fail on my Mac (M1 ARM chip, compiling sherpa natively) with current Sherpa main on Python 3.1o with numpy 1.21.4. Test failures persists after upgrading numpy to the current version 1.25. However, a look at the errors clearly shows that the problem is just numerics and not a real sherpa problem, the three failures are
```Python
def test_fit_results_str():
        """Just check we can call str() on a fit results instance"""
    
        data = make_data(Data1D)
        model = Const1D()
        fit = Fit(data, model, stat=Cash())
        fres = fit.fit()
        assert fres.succeeded
    
>       check_str(fres,
                  ["datasets       = None",
                   "itermethodname = none",
                   "methodname     = levmar",
                   "statname       = cash",
                   "succeeded      = True",
                   "parnames       = ('const1d.c0',)",
                   "parvals        = (3.33332926058462,)",
                   "statval        = -4.079456086503793",
                   "istatval       = 6.0",
                   "dstatval       = 10.079456086503793",
                   "numpoints      = 3",
                   "dof            = 2",
                   "qval           = None",
                   "rstat          = None",
                   "message        = successful termination",
                   "nfev           = 9"])

sherpa/tests/test_fit_unit.py:2850: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

out = <Fit results instance>
expecteds = ['datasets       = None', 'itermethodname = none', 'methodname     = levmar', 'statname       = cash', 'succeeded      = True', "parnames       = ('const1d.c0',)", ...]

    def check_str(out, expecteds):
        """Check that out, when split on a newline, matches expecteds"""
    
        toks = str(out).split("\n")
        for tok, expected in zip(toks, expecteds):
>           assert tok == expected
E           AssertionError: assert 'parvals     ...32926058504,)' == 'parvals     ...32926058462,)'
E             - parvals        = (3.33332926058462,)
E             ?                                 --
E             + parvals        = (3.33332926058504,)
E             ?                                ++

sherpa/tests/test_fit_unit.py:2836: AssertionError
```
and

```Python
./sherpa/tests/test_fit_unit.py::test_fit_results_with_iteration_str Failed: [undefined]AssertionError: assert 'parvals     ...00000003157,)' == 'parvals     ...00000000316,)'
  - parvals        = (3.250000000000316,)
  ?                                   ^
  + parvals        = (3.2500000000003157,)
  ?                                   ^^
def test_fit_results_with_iteration_str():
        """Just check we can call str() on a fit results instance with an iteration method
        """
    
        data = Data1D("x", [1, 2, 3, 4, 5], [2, 4, 12, 3, 4])
        data.staterror = [0.8] * 5
        model = Const1D()
        iter_opts = {'name': 'sigmarej', 'maxiters': 5, 'hrej': 5, 'lrej': 5, 'grow': 0}
        fit = Fit(data, model, stat=Chi2(), itermethod_opts=iter_opts)
        fres = fit.fit()
        assert fres.succeeded
    
>       check_str(fres,
                  ["datasets       = None",
                   "itermethodname = sigmarej",
                   "methodname     = levmar",
                   "statname       = chi2",
                   "succeeded      = True",
                   "parnames       = ('const1d.c0',)",
                   "parvals        = (3.250000000000316,)",
                   "statval        = 4.296875",
                   "istatval       = 225.0",
                   "dstatval       = 220.703125",
                   "numpoints      = 4",
                   "dof            = 3",
                   "qval           = 0.23114006377865534",
                   "rstat          = 1.4322916666666667",
                   "message        = successful termination",
                   "nfev           = 8"])

sherpa/tests/test_fit_unit.py:2907: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

out = <Fit results instance>
expecteds = ['datasets       = None', 'itermethodname = sigmarej', 'methodname     = levmar', 'statname       = chi2', 'succeeded      = True', "parnames       = ('const1d.c0',)", ...]

    def check_str(out, expecteds):
        """Check that out, when split on a newline, matches expecteds"""
    
        toks = str(out).split("\n")
        for tok, expected in zip(toks, expecteds):
>           assert tok == expected
E           AssertionError: assert 'parvals     ...00000003157,)' == 'parvals     ...00000000316,)'
E             - parvals        = (3.250000000000316,)
E             ?                                   ^
E             + parvals        = (3.2500000000003157,)
E             ?                                   ^^

sherpa/tests/test_fit_unit.py:2836: AssertionError
```
and

```Python
./sherpa/tests/test_fit_unit.py::test_esterrorresults_str Failed: [undefined]AssertionError: assert 'parvals     ...32926058504,)' == 'parvals     ...32926058462,)'
  - parvals     = (3.33332926058462,)
  ?                              --
  + parvals     = (3.33332926058504,)
  ?                             ++
def test_esterrorresults_str():
        """Basic check of str call."""
    
        data = make_data(Data1D)
        mdl = Const1D("mx")
        fit = Fit(data, mdl, stat=Cash())
        fit.fit()
    
        errs = fit.est_errors()
>       check_str(errs,
                  ["datasets    = None",
                   "methodname  = covariance",
                   "iterfitname = none",
                   "fitname     = levmar",
                   "statname    = cash",
                   "sigma       = 1",
                   "percent     = 68.26894921370858",
                   "parnames    = ('mx.c0',)",
                   "parvals     = (3.33332926058462,)",
                   "parmins     = (-1.0524866849982824,)",
                   "parmaxes    = (1.0524866849982824,)",
                   "nfits       = 0"])

sherpa/tests/test_fit_unit.py:3037: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

out = <covariance results instance>
expecteds = ['datasets    = None', 'methodname  = covariance', 'iterfitname = none', 'fitname     = levmar', 'statname    = cash', 'sigma       = 1', ...]

    def check_str(out, expecteds):
        """Check that out, when split on a newline, matches expecteds"""
    
        toks = str(out).split("\n")
        for tok, expected in zip(toks, expecteds):
>           assert tok == expected
E           AssertionError: assert 'parvals     ...32926058504,)' == 'parvals     ...32926058462,)'
E             - parvals     = (3.33332926058462,)
E             ?                              --
E             + parvals     = (3.33332926058504,)
E             ?                             ++

sherpa/tests/test_fit_unit.py:2836: AssertionError
```

--- a/sherpa/astro/ui/tests/test_astro_session.py
+++ b/sherpa/astro/ui/tests/test_astro_session.py
@@ -2527,17 +2527,7 @@ def test_save_delchi_data2d(session, emsg, idval, tmp_path):
             s.save_delchi(idval, str(out))
 
 
-def check_output(expected, got):
-    """Expected is broken down by lines, got is not"""
-
-    toks = got.split("\n")
-    for i, estr in enumerate(expected):
-        assert toks[i] == estr
-
-    assert len(toks) == len(expected)
-
-
-def check_save_ascii2d(session, expected, out, savefunc, idval, kwargs):
+def check_save_ascii2d(session, expected, out, savefunc, idval, kwargs, check_str):
     """Checking the output is tricky.
 
     If crates is in use with the AstroSession backend then the call
@@ -2567,14 +2557,14 @@ def check_save_ascii2d(session, expected, out, savefunc, idval, kwargs):
     else:
         savefunc(idval, str(out), **kwargs)
 
-    check_output(expected, out.read_text())
+    check_str(out.read_text(), expected)
 
 
 @pytest.mark.parametrize("session,kwargs,expected",
                          [(Session, {"comment": "!! "}, ["!! SOURCE", "7 11", ""]),
                           (AstroSession, {"ascii": True}, ["7", "11", ""])])
 @pytest.mark.parametrize("idval", [None, "bob"])
-def test_save_source_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io):
+def test_save_source_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io, check_str):
     """Basic check it works
 
     The output depends on the backend which is a bit distressing, and
@@ -2594,14 +2584,14 @@ def test_save_source_ascii_data2d(session, kwargs, expected, idval, tmp_path, sk
     s.set_source(idval, mdl)
 
     out = tmp_path / "created.dat"
-    check_save_ascii2d(session, expected, out, s.save_source, idval, kwargs)
+    check_save_ascii2d(session, expected, out, s.save_source, idval, kwargs, check_str)
 
 
 @pytest.mark.parametrize("session,kwargs,expected",
                          [(Session, {"comment": ""}, ["MODEL", "7 11", ""]),
                           (AstroSession, {"ascii": True}, ["7", "11", ""])])
 @pytest.mark.parametrize("idval", [None, "bob"])
-def test_save_model_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io):
+def test_save_model_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io, check_str):
     """Basic check it works
 
     The output depends on the backend which is a bit distressing, and
@@ -2621,14 +2611,14 @@ def test_save_model_ascii_data2d(session, kwargs, expected, idval, tmp_path, ski
     s.set_source(idval, mdl)
 
     out = tmp_path / "created.dat"
-    check_save_ascii2d(session, expected, out, s.save_model, idval, kwargs)
+    check_save_ascii2d(session, expected, out, s.save_model, idval, kwargs, check_str)
 
 
 @pytest.mark.parametrize("session,kwargs,expected",
                          [(Session, {}, ["#RESID", "-7 -11", ""]),
                           (AstroSession, {"ascii": True}, ["-7", "-11", ""])])
 @pytest.mark.parametrize("idval", [None, "bob"])
-def test_save_resid_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io):
+def test_save_resid_ascii_data2d(session, kwargs, expected, idval, tmp_path, skip_if_no_io, check_str):
     """Basic check it works
 
     The output depends on the backend which is a bit distressing.
@@ -2645,7 +2635,7 @@ def test_save_resid_ascii_data2d(session, kwargs, expected, idval, tmp_path, ski
     s.set_source(idval, mdl)
 
     out = tmp_path / "created.dat"
-    check_save_ascii2d(session, expected, out, s.save_resid, idval, kwargs)
+    check_save_ascii2d(session, expected, out, s.save_resid, idval, kwargs, check_str)
 
 
 @pytest.mark.parametrize("session", [Session, AstroSession])

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -696,7 +696,7 @@ eterm = r"(?:e[+-]?\d+)"
 term = "|".join([fr"[+-]?\d+\.\d*{eterm}?",
                  fr"[+-]?\.\d+{eterm}?"
                  fr"[+-]?\d+{eterm}"])
-PATTERN = re.compile(f"(.*)(?<![\d+-])({term})")
+PATTERN = re.compile(fr"(.*)(?<![\d+-])({term})")
 
 
 class NumberChecker:

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -689,8 +689,8 @@ def add_sherpa_test_data_dir(doctest_namespace):
 
 
 # Try to come up with a regexp for a numeric literal.  This is
-# informed by pytest-doctestlpus but this is a lot easier as we only
-# care about a single value.
+# informed by pytest-doctestplus but this is a lot easier as we only
+# care about a single number.
 #
 eterm = r"(?:e[+-]?\d+)"
 term = "|".join([fr"[+-]?\d+\.\d*{eterm}?",
@@ -702,19 +702,14 @@ PATTERN = re.compile(f"(.*)(?<![\d+-])({term})")
 class NumberChecker:
     """Allow a string to be compared, allowing for tolerances.
 
-    This is written to match the re.Pattern interface as used by
-    check_str_fixture (match routine and pattern attribute). It is a
-    limited version of
-    pytest_doctestplus.output_checker.OutputChecker.
-
-    It only checks a single number value in the text. This could be
-    updated to be closer to doctestplus if found to be useful.
+    It is a limited version of
+    pytest_doctestplus.output_checker.OutputChecker: it only checks a
+    single number value in the text. This could be updated to be
+    closer to doctestplus if found to be useful.
 
     """
 
     def __init__(self, expected, rtol=1e-4, atol=None):
-        self.pattern = expected
-
         lhs, value, rhs = self.split(expected)
         self.lhs = lhs
         self.number = value
@@ -734,7 +729,10 @@ class NumberChecker:
 
         match = PATTERN.match(text)
         if match is None:
-            raise ValueError(f"No number found in '{text}'")
+            # Make sure that we note a case where the regexp has
+            # failed to find a number.
+            #
+            raise ValueError(f"Error in test - no number found in '{text}'")
 
         lhs = match[1]
         number = float(match[2])
@@ -742,20 +740,13 @@ class NumberChecker:
 
         return lhs, number, rhs
 
-    def match(self, got):
-        """This abuses the match API"""
+    def check(self, got):
+        """check that got matches the expected text"""
         lhs, value, rhs = self.split(got)
         assert lhs == self.lhs
         assert rhs == self.rhs
         assert value == pytest.approx(self.number,
                                       rel=self.rtol, abs=self.atol)
-        return True
-
-
-@pytest.fixture
-def check_number():
-    """Convert a string containing a number into a NumberChecker"""
-    return NumberChecker
 
 
 def check_str_fixture(out, expecteds):
@@ -766,24 +757,41 @@ def check_str_fixture(out, expecteds):
     (to avoid the need to mark up expected text as a regexp) but this
     is problematic at the moment, so it is not done. For now we
     require users to explictly mark up those lines which we want to
-    check with numeric constraints. Maybe we should do this for all
-    lines (at least, those not already done as regexps) but let's see
-    how this works.
+    check with numeric constraints by adding the text (taken from
+    doctestplus)
+
+        "# doctest: +FLOAT_CMP"
+
+    to the end of each line that needs it.
 
     Parameters
     ----------
     out : str
-        The output to check
+        The output to check (each line is compared to the expecteds
+        array).
     expecteds : list of str or Pattern
-        The expected output, split on new-line.
+        The expected output, split on new-line. Each line is one of: a
+        Pattern, a string ending in "# doctest: +FLOAT_CMP", or a
+        string (which is just checked for equality).
 
     """
 
     toks = str(out).split("\n")
     for tok, expected in zip(toks, expecteds):
+        # expected is one of:
+        #  - a pattern
+        #  - a string ending in #doctest: +FLOAT_CMP
+        #  - a normal string
+        #
         try:
             assert expected.match(tok), (expected.pattern, tok)
         except AttributeError:
+            idx = expected.find("# doctest: +FLOAT_CMP")
+            if idx > -1:
+                pat = NumberChecker(expected[:idx].rstrip())
+                pat.check(tok)
+                continue
+
             assert tok == expected
 
     assert len(toks) == len(expecteds)

--- a/sherpa/conftest.py
+++ b/sherpa/conftest.py
@@ -661,6 +661,7 @@ def cleanup_ds9_backend():
     from sherpa.image import backend
     backend.close()
 
+
 @pytest.fixture(autouse=True)
 def add_sherpa_test_data_dir(doctest_namespace):
     '''Define `data_dir` for doctests
@@ -685,3 +686,116 @@ def add_sherpa_test_data_dir(doctest_namespace):
         path += '/'
 
     doctest_namespace["data_dir"] = path
+
+
+# Try to come up with a regexp for a numeric literal.  This is
+# informed by pytest-doctestlpus but this is a lot easier as we only
+# care about a single value.
+#
+eterm = r"(?:e[+-]?\d+)"
+term = "|".join([fr"[+-]?\d+\.\d*{eterm}?",
+                 fr"[+-]?\.\d+{eterm}?"
+                 fr"[+-]?\d+{eterm}"])
+PATTERN = re.compile(f"(.*)(?<![\d+-])({term})")
+
+
+class NumberChecker:
+    """Allow a string to be compared, allowing for tolerances.
+
+    This is written to match the re.Pattern interface as used by
+    check_str_fixture (match routine and pattern attribute). It is a
+    limited version of
+    pytest_doctestplus.output_checker.OutputChecker.
+
+    It only checks a single number value in the text. This could be
+    updated to be closer to doctestplus if found to be useful.
+
+    """
+
+    def __init__(self, expected, rtol=1e-4, atol=None):
+        self.pattern = expected
+
+        lhs, value, rhs = self.split(expected)
+        self.lhs = lhs
+        self.number = value
+        self.rhs = rhs
+
+        self.rtol = rtol
+        self.atol = atol
+
+    def split(self, text):
+        """Split a string on a number
+
+        Parameters
+        ----------
+        text : str
+            The string containing a single number.
+        """
+
+        match = PATTERN.match(text)
+        if match is None:
+            raise ValueError(f"No number found in '{text}'")
+
+        lhs = match[1]
+        number = float(match[2])
+        rhs = text[match.end():]
+
+        return lhs, number, rhs
+
+    def match(self, got):
+        """This abuses the match API"""
+        lhs, value, rhs = self.split(got)
+        assert lhs == self.lhs
+        assert rhs == self.rhs
+        assert value == pytest.approx(self.number,
+                                      rel=self.rtol, abs=self.atol)
+        return True
+
+
+@pytest.fixture
+def check_number():
+    """Convert a string containing a number into a NumberChecker"""
+    return NumberChecker
+
+
+def check_str_fixture(out, expecteds):
+    """Check that out, when split on a newline, matches expecteds.
+
+    It would be nice to use use doctestplus - that is
+    pytest_doctestplus.output_checker.OutputChecker - for the checking
+    (to avoid the need to mark up expected text as a regexp) but this
+    is problematic at the moment, so it is not done. For now we
+    require users to explictly mark up those lines which we want to
+    check with numeric constraints. Maybe we should do this for all
+    lines (at least, those not already done as regexps) but let's see
+    how this works.
+
+    Parameters
+    ----------
+    out : str
+        The output to check
+    expecteds : list of str or Pattern
+        The expected output, split on new-line.
+
+    """
+
+    toks = str(out).split("\n")
+    for tok, expected in zip(toks, expecteds):
+        try:
+            assert expected.match(tok), (expected.pattern, tok)
+        except AttributeError:
+            assert tok == expected
+
+    assert len(toks) == len(expecteds)
+
+
+@pytest.fixture
+def check_str():
+    """Returns the check_str_fixture routine.
+
+    See check_str_fixture for documentation. This is a bit of a hack
+    to treat a fixture as a way of importing check_str_fixture.
+
+    """
+
+    return check_str_fixture

--- a/sherpa/estmethods/tests/test_estmethods.py
+++ b/sherpa/estmethods/tests/test_estmethods.py
@@ -223,71 +223,52 @@ def test_estmethod_repr(cls, name):
     assert repr(m) == f"<{name} error-estimation method instance '{name.lower()}'>"
 
 
-def check_output(out, expecteds):
-    """Check out (str) matches expecteds after splitting newlines
-
-    There's a special check for the numcores line, as the answer
-    depends on the number of cores present, so we drop that part
-    of the check.
-
-    """
-    toks = out.split("\n")
-    for tok, expected in zip(toks, expecteds):
-
-        if expected is None:
-            assert re.match(r"^numcores     ?= \d+$", tok)
-        else:
-            assert tok == expected
-
-    assert len(toks) == len(expecteds)
-
-
-def test_estmethod_str_covariance():
+def test_estmethod_str_covariance(check_str):
     """Simple check."""
     m = Covariance()
-    check_output(str(m),
-                 ["name        = covariance",
-                  "sigma       = 1",
-                  "eps         = 0.01",
-                  "maxiters    = 200",
-                  "soft_limits = False"])
+    check_str(str(m),
+              ["name        = covariance",
+               "sigma       = 1",
+               "eps         = 0.01",
+               "maxiters    = 200",
+               "soft_limits = False"])
 
 
-def test_estmethod_str_confidence():
+def test_estmethod_str_confidence(check_str):
     """Simple check."""
     m = Confidence()
-    check_output(str(m),
-                 ["name         = confidence",
-                  "sigma        = 1",
-                  "eps          = 0.01",
-                  "maxiters     = 200",
-                  "soft_limits  = False",
-                  "remin        = 0.01",
-                  "fast         = False",
-                  "parallel     = True",
-                  None,  # special-case numcores line
-                  "maxfits      = 5",
-                  "max_rstat    = 3",
-                  "tol          = 0.2",
-                  "verbose      = False",
-                  "openinterval = False"
-                  ])
+    check_str(str(m),
+              ["name         = confidence",
+               "sigma        = 1",
+               "eps          = 0.01",
+               "maxiters     = 200",
+               "soft_limits  = False",
+               "remin        = 0.01",
+               "fast         = False",
+               "parallel     = True",
+               re.compile(r"^numcores     ?= \d+$"),
+               "maxfits      = 5",
+               "max_rstat    = 3",
+               "tol          = 0.2",
+               "verbose      = False",
+               "openinterval = False"
+               ])
 
 
-def test_estmethod_str_projection():
+def test_estmethod_str_projection(check_str):
     """Simple check."""
     m = Projection()
-    check_output(str(m),
-                 ["name        = projection",
-                  "sigma       = 1",
-                  "eps         = 0.01",
-                  "maxiters    = 200",
-                  "soft_limits = False",
-                  "remin       = 0.01",
-                  "fast        = False",
-                  "parallel    = True",
-                  None,  # special-case numcores line
-                  "maxfits     = 5",
-                  "max_rstat   = 3",
-                  "tol         = 0.2"
-                  ])
+    check_str(str(m),
+              ["name        = projection",
+               "sigma       = 1",
+               "eps         = 0.01",
+               "maxiters    = 200",
+               "soft_limits = False",
+               "remin       = 0.01",
+               "fast        = False",
+               "parallel    = True",
+               re.compile(r"^numcores     ?= \d+$"),
+               "maxfits     = 5",
+               "max_rstat   = 3",
+               "tol         = 0.2"
+               ])

--- a/sherpa/sim/tests/test_asymmetric.py
+++ b/sherpa/sim/tests/test_asymmetric.py
@@ -272,7 +272,7 @@ def test_zero_case(reset_seed):
     assert (stats >= bestfit.statval).all()
 
 
-def test_resample_supports_data1d(caplog, reset_seed):
+def test_resample_supports_data1d(caplog, reset_seed, check_str):
 
     orig = Data1D("orig", [1, 2, 3], [4, 2, 5], [0.1, 0.2, 0.5])
     model = Const1D("mdl")
@@ -285,7 +285,8 @@ def test_resample_supports_data1d(caplog, reset_seed):
     lname, lvl, msg = caplog.record_tuples[0]
     assert lname == "sherpa"
     assert lvl == logging.INFO
-    assert msg == "mdl.c0 : avg = 1.0343987745935705 , std = 0.5208696279243179"
+    check_str(msg,
+              ["mdl.c0 : avg = 1.0343987745935705 , std = 0.5208696279243179  # doctest: +FLOAT_CMP"])
 
 
 def test_resample_fails_unsupported_data():

--- a/sherpa/tests/test_fit_unit.py
+++ b/sherpa/tests/test_fit_unit.py
@@ -2828,7 +2828,7 @@ def test_fit_ensures_data_and_model_dimensionality_matches(data_class, ddim, mod
         Fit(data, model)
 
 
-def test_fit_results_str(check_str, check_number):
+def test_fit_results_str(check_str):
     """Just check we can call str() on a fit results instance"""
 
     data = make_data(Data1D)
@@ -2844,10 +2844,10 @@ def test_fit_results_str(check_str, check_number):
                "statname       = cash",
                "succeeded      = True",
                "parnames       = ('const1d.c0',)",
-               check_number("parvals        = (3.33332926058462,)"),
-               check_number("statval        = -4.079456086503793"),
+               "parvals        = (3.33332926058462,)  # doctest: +FLOAT_CMP",
+               "statval        = -4.079456086503793   # doctest: +FLOAT_CMP",
                "istatval       = 6.0",
-               check_number("dstatval       = 10.079456086503793"),
+               "dstatval       = 10.079456086503793   # doctest: +FLOAT_CMP",
                "numpoints      = 3",
                "dof            = 2",
                "qval           = None",
@@ -2882,7 +2882,7 @@ def test_fit_itermethod_requires_known_name():
         Fit(data, model, itermethod_opts={"name": "not-a-name", "grow": 2})
 
 
-def test_fit_results_with_iteration_str(check_str, check_number):
+def test_fit_results_with_iteration_str(check_str):
     """Just check we can call str() on a fit results instance with an iteration method
     """
 
@@ -2901,14 +2901,14 @@ def test_fit_results_with_iteration_str(check_str, check_number):
                "statname       = chi2",
                "succeeded      = True",
                "parnames       = ('const1d.c0',)",
-               check_number("parvals        = (3.250000000000316,)"),
-               check_number("statval        = 4.296875"),
+               "parvals        = (3.250000000000316,)  # doctest: +FLOAT_CMP",
+               "statval        = 4.296875              # doctest: +FLOAT_CMP",
                "istatval       = 225.0",
-               check_number("dstatval       = 220.703125"),
+               "dstatval       = 220.703125            # doctest: +FLOAT_CMP",
                "numpoints      = 4",
                "dof            = 3",
-               check_number("qval           = 0.23114006377865534"),
-               check_number("rstat          = 1.4322916666666667"),
+               "qval           = 0.23114006377865534   # doctest: +FLOAT_CMP",
+               "rstat          = 1.4322916666666667    # doctest: +FLOAT_CMP",
                "message        = successful termination",
                "nfev           = 8"])
 
@@ -3015,7 +3015,7 @@ def test_fit_simulfit_multiple_single_model():
     assert fres.parvals == pytest.approx([3.8333336041446615])
 
 
-def test_esterrorresults_str(check_str, check_number):
+def test_esterrorresults_str(check_str):
     """Basic check of str call."""
 
     data = make_data(Data1D)
@@ -3033,9 +3033,9 @@ def test_esterrorresults_str(check_str, check_number):
                "sigma       = 1",
                "percent     = 68.26894921370858",
                "parnames    = ('mx.c0',)",
-               check_number("parvals     = (3.33332926058462,)"),
-               check_number("parmins     = (-1.0524866849982824,)"),
-               check_number("parmaxes    = (1.0524866849982824,)"),
+               "parvals     = (3.33332926058462,)     # doctest: +FLOAT_CMP",
+               "parmins     = (-1.0524866849982824,)  # doctest: +FLOAT_CMP",
+               "parmaxes    = (1.0524866849982824,)   # doctest: +FLOAT_CMP",
                "nfits       = 0"])
 
 

--- a/sherpa/tests/test_fit_unit.py
+++ b/sherpa/tests/test_fit_unit.py
@@ -2828,17 +2828,7 @@ def test_fit_ensures_data_and_model_dimensionality_matches(data_class, ddim, mod
         Fit(data, model)
 
 
-def check_str(out, expecteds):
-    """Check that out, when split on a newline, matches expecteds"""
-
-    toks = str(out).split("\n")
-    for tok, expected in zip(toks, expecteds):
-        assert tok[:28] == expected[:28]
-
-    assert len(toks) == len(expecteds)
-
-
-def test_fit_results_str():
+def test_fit_results_str(check_str, check_number):
     """Just check we can call str() on a fit results instance"""
 
     data = make_data(Data1D)
@@ -2854,10 +2844,10 @@ def test_fit_results_str():
                "statname       = cash",
                "succeeded      = True",
                "parnames       = ('const1d.c0',)",
-               "parvals        = (3.33332926058462,)",
-               "statval        = -4.079456086503793",
+               check_number("parvals        = (3.33332926058462,)"),
+               check_number("statval        = -4.079456086503793"),
                "istatval       = 6.0",
-               "dstatval       = 10.079456086503793",
+               check_number("dstatval       = 10.079456086503793"),
                "numpoints      = 3",
                "dof            = 2",
                "qval           = None",
@@ -2892,7 +2882,7 @@ def test_fit_itermethod_requires_known_name():
         Fit(data, model, itermethod_opts={"name": "not-a-name", "grow": 2})
 
 
-def test_fit_results_with_iteration_str():
+def test_fit_results_with_iteration_str(check_str, check_number):
     """Just check we can call str() on a fit results instance with an iteration method
     """
 
@@ -2911,14 +2901,14 @@ def test_fit_results_with_iteration_str():
                "statname       = chi2",
                "succeeded      = True",
                "parnames       = ('const1d.c0',)",
-               "parvals        = (3.250000000000316,)",
-               "statval        = 4.296875",
+               check_number("parvals        = (3.250000000000316,)"),
+               check_number("statval        = 4.296875"),
                "istatval       = 225.0",
-               "dstatval       = 220.703125",
+               check_number("dstatval       = 220.703125"),
                "numpoints      = 4",
                "dof            = 3",
-               "qval           = 0.23114006377865534",
-               "rstat          = 1.4322916666666667",
+               check_number("qval           = 0.23114006377865534"),
+               check_number("rstat          = 1.4322916666666667"),
                "message        = successful termination",
                "nfev           = 8"])
 
@@ -3025,7 +3015,7 @@ def test_fit_simulfit_multiple_single_model():
     assert fres.parvals == pytest.approx([3.8333336041446615])
 
 
-def test_esterrorresults_str():
+def test_esterrorresults_str(check_str, check_number):
     """Basic check of str call."""
 
     data = make_data(Data1D)
@@ -3043,13 +3033,13 @@ def test_esterrorresults_str():
                "sigma       = 1",
                "percent     = 68.26894921370858",
                "parnames    = ('mx.c0',)",
-               "parvals     = (3.33332926058462,)",
-               "parmins     = (-1.0524866849982824,)",
-               "parmaxes    = (1.0524866849982824,)",
+               check_number("parvals     = (3.33332926058462,)"),
+               check_number("parmins     = (-1.0524866849982824,)"),
+               check_number("parmaxes    = (1.0524866849982824,)"),
                "nfits       = 0"])
 
 
-def test_fit_outfile_simple_check(tmp_path):
+def test_fit_outfile_simple_check(tmp_path, check_str):
 
     outfile = tmp_path / "sherpa.save"
 

--- a/sherpa/tests/test_fit_unit.py
+++ b/sherpa/tests/test_fit_unit.py
@@ -2833,7 +2833,7 @@ def check_str(out, expecteds):
 
     toks = str(out).split("\n")
     for tok, expected in zip(toks, expecteds):
-        assert tok == expected
+        assert tok[:28] == expected[:28]
 
     assert len(toks) == len(expecteds)
 

--- a/sherpa/tests/test_tests.py
+++ b/sherpa/tests/test_tests.py
@@ -18,15 +18,23 @@
 '''Make sure we tests the test fixtures.'''
 import pytest
 
-def test_check_str(check_str):
+def test_check_str_fails(check_str):
     '''The tests that use check_str make sure that it passes,
-    but what if it were to pass everything? 
+    but what if it were to pass everything?
     So, add a test here to make sure it actually fails when it should.'''
-    with pytest.raises(AssertionError):
-        check_str('foo = 1.23456', 'foo = 1.230000')
+    with pytest.raises(AssertionError, match="^assert 'foo = "):
+        check_str('foo = 1.23456', ['foo = 1.230000'])
 
-    with pytest.raises(AssertionError):
-        check_str('foo = 1.23456', 'foo = 1.230000   # doctest: +FLOAT_CMP')
 
-    with pytest.raises(AssertionError):
-        check_str('1.23456', '1.230000   # doctest: +FLOAT_CMP')
+def test_check_str_fails_with_float_cmp(check_str):
+    # The match value may need updating if pytest.approx changes it's
+    # assertion error.
+    with pytest.raises(AssertionError, match="^assert 1.23456 "):
+        check_str('foo = 1.23456', ['foo = 1.230000   # doctest: +FLOAT_CMP'])
+
+
+def test_check_str_fails_with_float_cmp_number_only(check_str):
+    # The match value may need updating if pytest.approx changes it's
+    # assertion error.
+    with pytest.raises(AssertionError, match="^assert 1.23456 "):
+        check_str('1.23456', ['1.230000   # doctest: +FLOAT_CMP  '])

--- a/sherpa/tests/test_tests.py
+++ b/sherpa/tests/test_tests.py
@@ -1,0 +1,32 @@
+#
+#  Copyright (C) 2023  MIT
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+'''Make sure we tests the test fixtures.'''
+import pytest
+
+def test_check_str(check_str):
+    '''The tests that use check_str make sure that it passes,
+    but what if it were to pass everything? 
+    So, add a test here to make sure it actually fails when it should.'''
+    with pytest.raises(AssertionError):
+        check_str('foo = 1.23456', 'foo = 1.230000')
+
+    with pytest.raises(AssertionError):
+        check_str('foo = 1.23456', 'foo = 1.230000   # doctest: +FLOAT_CMP')
+
+    with pytest.raises(AssertionError):
+        check_str('1.23456', '1.230000   # doctest: +FLOAT_CMP')

--- a/sherpa/ui/tests/test_ui_unit.py
+++ b/sherpa/ui/tests/test_ui_unit.py
@@ -1189,17 +1189,7 @@ def test_set_filter_masked_wrong(clean_ui):
         ui.set_filter(np.asarray([True, False]))
 
 
-def check_str(out, expecteds):
-    """Check that out, when split on a newline, matches expecteds"""
-
-    toks = str(out).split("\n")
-    for tok, expected in zip(toks, expecteds):
-        assert tok == expected
-
-    assert len(toks) == len(expecteds)
-
-
-def test_fit_output_single(clean_ui, caplog):
+def test_fit_output_single(clean_ui, caplog, check_str):
     """This is essentially the test from #1804 but for a single dataset"""
 
     ui.load_arrays(1, [1, 2, 3], [12, 10, 4])
@@ -1225,7 +1215,7 @@ def test_fit_output_single(clean_ui, caplog):
                "   mdl.c0         8.66665      +/- 1.70862     "])
 
 
-def test_calc_stat_info_output_single(clean_ui, caplog):
+def test_calc_stat_info_output_single(clean_ui, caplog, check_str):
     """This is essentially the test from #1804 but for a single dataset
     and converted to check calc_stat
 
@@ -1252,7 +1242,7 @@ def test_calc_stat_info_output_single(clean_ui, caplog):
                "Degrees of freedom    = 2"])
 
 
-def test_fit_output_multi(clean_ui, caplog):
+def test_fit_output_multi(clean_ui, caplog, check_str):
     """This is essentially the test from #1804"""
 
     ui.load_arrays(1, [1, 2, 3], [12, 10, 4])
@@ -1280,7 +1270,7 @@ def test_fit_output_multi(clean_ui, caplog):
                "   mdl.c0         7.83332      +/- 1.14642     "])
 
 
-def test_calc_stat_info_output_multi(clean_ui, caplog):
+def test_calc_stat_info_output_multi(clean_ui, caplog, check_str):
     """This is essentially the test from #1804 but for calc_stat_info"""
 
     ui.load_arrays(1, [1, 2, 3], [12, 10, 4])

